plugins/experimental/icap/icap_plugin.cc

```diff
@@ -0,0 +1,1267 @@
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+#include <string>
+#include <regex>
+
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <iostream>
+#include <sstream>
+
+#include "ts/ts.h"
+
+#define PLUGIN_NAME "icap_plugin"
+
+#define STATE_BEGIN 1
+#define STATE_CONNECT 2
+#define STATE_WRITE_HEADER 3
+#define STATE_WRITE_BODY 4
+#define STATE_READ_ICAP_HEADER 5
+#define STATE_READ_HTTP_HEADER 6
+#define STATE_READ_HTTP_BODY 7
+#define STATE_SEND_ERROR_MSG 8
+#define STATE_BYPASS 9
+#define STATE_BUFFER_OS_RESP 10
+#define STATE_SEND_OS_RESP 11
+
+#define ICAP_SERVICE_URL "icap://127.0.0.1/avscan"
+#define ICAP_VERSION "1.0"
+
+struct TransformData {
+  int state;
+  TSHttpTxn txn;
+
+  int64_t server_reply_content_length;
+
+  TSIOBuffer input_buf;
+  TSIOBufferReader input_reader;
+
+  TSIOBuffer os_resp_buf;
+  TSIOBufferReader os_resp_reader;
+
+  int64_t done_write;
+
+  TSIOBuffer icap_resp_buf;
+  TSIOBufferReader icap_resp_reader;
+
+  TSIOBuffer output_buf;
+  TSIOBufferReader output_reader;
+  TSVConn output_vc;
+  TSVIO output_vio;
+
+  TSAction pending_action;
+  TSVConn icap_vc;
+  TSVIO icap_vio;
+
+  std::string icap_header;
+  std::string http_header;
+  std::string chunk_length_str;
+  int64_t icap_reply_content_length;
+
+  int64_t http_body_chunk_length;
+  int64_t http_body_total_length_written;
+
+  bool eos_detected;
+
+  std::string err_msg;
+
+  TransformData(TSHttpTxn txnp);
+  ~TransformData();
+};
+
+/* Configurable paremeteres */
+static std::string server_ip;
+static int server_port;
+static int carp_port;
+static int debug_enabled;
+
+/* Stats for debug */
+static int scan_passed;
+static int scan_failed;
+static int icap_conn_failed;
+static int total_icap_invalid;
+static int icap_response_err;
+static int icap_write_failed;
+
+static int transform_handler(TSCont contp, TSEvent event, void *edata);
+static int transform_read_http_header_event(TSCont contp, TransformData *data, TSEvent event, void *edata);
+static int transform_send_error_msg(TSCont contp, TransformData *data);
+static int transform_bypass(TSCont contp, TransformData *data);
+static int transform_send_os_resp(TSCont contp, TransformData *data);
+
+TransformData::TransformData(TSHttpTxn txnp)
```

Comment:
Why not:

TransformData::TransformData() : txn(txnp) {}

And initialize the rest in the class definition.  Also, txn could be const.